zookeeper-server/src/main/java/org/apache/zookeeper/server/ReferenceCountedACLCache.java

```diff
@@ -125,10 +125,14 @@ public synchronized void deserialize(InputArchive ia) throws IOException {
         }
     }
 
-    public synchronized void serialize(OutputArchive oa) throws IOException {
-        oa.writeInt(longKeyMap.size(), "map");
-        Set<Map.Entry<Long, List<ACL>>> set = longKeyMap.entrySet();
-        for (Map.Entry<Long, List<ACL>> val : set) {
+    public void serialize(OutputArchive oa) throws IOException {
+        Map<Long, List<ACL>> clonedLongKeyMap;
+        synchronized (this)
+        {
+            clonedLongKeyMap = new HashMap<Long, List<ACL>>(longKeyMap);
+        }
+        oa.writeInt(clonedLongKeyMap.size(), "map");
```

Comment:
synchronized (this) {
            clonedLongKeyMap = new HashMap<>(longKeyMap);
        }